File: client/ios/safeme/Features/ViewModels/IncidentsViewModel.swift
import Foundation
import Combine

struct CreateIncidentRequest: Encodable {
    let type: String
    let description: String
    let lat: Double
    let lng: Double
    let reporter_hash: String
    let city: String
    let plate: String?
}

struct IncidentResponse: Decodable, Identifiable {
    let id: String
    let type: String
    let description: String
    let lat: Double
    let lng: Double
    let status: String
    let timestamp: String
}

@MainActor
class IncidentsViewModel: ObservableObject {
    @Published var nearbyIncidents: [IncidentResponse] = []
    @Published var isSubmitting = false
    @Published var submissionSuccess = false
    
    func fetchNearbyIncidents(lat: Double, lng: Double) async {
        do {
            let endpoint = APIEndpoint(
                path: "/incidents/nearby",
                queryItems: [
                    URLQueryItem(name: "lat", value: String(lat)),
                    URLQueryItem(name: "lng", value: String(lng))
                ]
            )
            nearbyIncidents = try await APIClient.shared.request(endpoint, responseType: [IncidentResponse].self)
        } catch {
            print("Error fetching incidents: \(error)")
        }
    }
    
    func reportIncident(type: String, description: String, lat: Double, lng: Double, city: String, plate: String? = nil) async {
        isSubmitting = true
        submissionSuccess = false
        defer { isSubmitting = false }
        
        do {
            let request = CreateIncidentRequest(
                type: type,
                description: description,
                lat: lat,
                lng: lng,
                reporter_hash: "simulated_reporter_hash_ios",
                city: city,
                plate: plate
            )
            let endpoint = APIEndpoint(path: "/incidents", method: "POST", body: try? JSONEncoder().encode(request))
            _ = try await APIClient.shared.request(endpoint, responseType: IncidentResponse.self)
            
            submissionSuccess = true
        } catch {
            print("Error reporting incident: \(error)")
        }
    }
}

File: client/android/app/src/main/java/com/solutioncrafts/safeme/ui/features/viewmodels/IncidentsViewModel.kt
package com.solutioncrafts.safeme.ui.features.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.solutioncrafts.safeme.data.network.CreateIncidentRequest
import com.solutioncrafts.safeme.data.network.IncidentResponse
import com.solutioncrafts.safeme.data.network.NetworkClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

class IncidentsViewModel : ViewModel() {
    private val _nearbyIncidents = MutableStateFlow<List<IncidentResponse>>(emptyList())
    val nearbyIncidents: StateFlow<List<IncidentResponse>> = _nearbyIncidents

    private val _isSubmitting = MutableStateFlow(false)
    val isSubmitting: StateFlow<Boolean> = _isSubmitting

    private val _submissionSuccess = MutableStateFlow(false)
    val submissionSuccess: StateFlow<Boolean> = _submissionSuccess

    fun fetchNearbyIncidents(lat: Double, lng: Double) {
        viewModelScope.launch {
            try {
                val incidents = NetworkClient.apiService.getNearbyIncidents(lat, lng)
                _nearbyIncidents.value = incidents
            } catch (e: Exception) {
                // handle error
            }
        }
    }

    fun reportIncident(
        type: String,
        description: String,
        lat: Double,
        lng: Double,
        city: String,
        plate: String? = null
    ) {
        viewModelScope.launch {
            _isSubmitting.value = true
            _submissionSuccess.value = false
            try {
                val request = CreateIncidentRequest(
                    type = type,
                    description = description,
                    lat = lat,
                    lng = lng,
                    reporter_hash = "simulated_reporter_hash",
                    city = city,
                    plate = plate
                )
                NetworkClient.apiService.createIncident(request)
                _submissionSuccess.value = true
            } catch (e: Exception) {
                // handle error
            } finally {
                _isSubmitting.value = false
            }
        }
    }
}
